### PR TITLE
cmd/stunstamp: implement service to measure DERP STUN RTT

### PR DESCRIPTION
**Pull request by jwhited**
stunstamp timestamping includes userspace and SO_TIMESTAMPING kernel timestamping where available. Measurements are written locally to a sqlite DB and exposed over an HTTP API.

Updates tailscale/corp#20344

### CODE REVIEW

**Comment by knyar** [comment]
I am curious what the main motivation is for having a bespoke time series database here instead of exporting metrics to our Prometheus server?

If I understood this correctly, we are running 8 probes per DERP server, once every 10-60 seconds. If we export 2 counters (cumulative number of probes, cumulative rtt) per probe, it will result in 1184 new time series (0.3% increase to our [existing 370K](http://monitoring:9090/tsdb-status)). Our default collection interval is 5s, so even with a 10s probe interval, metrics will be collected after every probe.

File: cmd/stunstamp/stunstamp.go [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+			conns[timestampSourceUserspace].Close()
+			delete(stableConns, k)
+		}
+	}
+
+	for {
+		select {
+		case err := <-errCh:
+			close(doneCh)
+			wg.Wait()
```
Comment (knyar) [inline cmd/stunstamp/stunstamp.go:311]: I've not found any `wg.Done()` calls on this, is it expected?

File: cmd/stunstamp/stunstamp.go [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+		log.Fatalf("retention must be >= interval")
+	}
+	if len(*flagRemoteWriteURL) < 1 {
+		log.Fatalf("rw-url flag is unset")
+	}
+	_, err := url.Parse(*flagRemoteWriteURL)
+	if err != nil {
+		log.Fatalf("invalid rw-url flag value: %v", err)
+	}
+	if len(*flagInstance) < 1 {
```
Comment (knyar) [inline cmd/stunstamp/stunstamp.go:478]: Perhaps default to `os.Hostname()`?

File: cmd/stunstamp/stunstamp.go [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+					wg.Add(1)
+					numProbes += 2
+					go doProbe(v6StableConns[timestampSourceKernel], meta6, timestampSourceKernel)
+					go doProbe(nil, meta6, timestampSourceKernel)
+				}
+			}
+		}
+	}
+
+	// cleanup conns we no longer need
```
Comment (knyar) [inline cmd/stunstamp/stunstamp.go:296]: Prometheus remote write protocol also expects [stale markers](https://prometheus.io/docs/concepts/remote_write_spec/#stale-markers) to be written for time series that will no longer be reported.

Perhaps we could store labelsets (`[]prompb.Label`) per probed address, and have a similar cleanup check that would write stale markers every time a DERP address disappears? Would also make sense to try to write them at shutdown time.

File: cmd/stunstamp/stunstamp.go (hunk 1 above)
Comment (jwhited) [reply cmd/stunstamp/stunstamp.go:312]: updated in https://github.com/tailscale/tailscale/pull/12241/commits/41c045be5ffcb84ebe6b3d0794baff17689236e2

File: cmd/stunstamp/stunstamp.go (hunk 2 above)
Comment (jwhited) [reply cmd/stunstamp/stunstamp.go:479]: updated in https://github.com/tailscale/tailscale/pull/12241/commits/3876f85d957d06a89174247d21544f705be42c10

File: cmd/stunstamp/stunstamp.go (hunk 3 above)
Comment (jwhited) [reply cmd/stunstamp/stunstamp.go:320]: updated in https://github.com/tailscale/tailscale/pull/12241/commits/1381c4675e3bc521725856e6d605fbf55073d6ad

**Review by knyar — APPROVED** [review]

File: cmd/stunstamp/stunstamp.go [hunk 4]
```diff
@@ -0,0 +1,778 @@
...
+	uconn, err := net.ListenUDP("udp", &net.UDPAddr{})
+	if err != nil {
+		return conns, err
+	}
+	conns[timestampSourceUserspace] = uconn
+	stableConns[addr] = conns
+	return conns, nil
+}
+
+func probeNodes(nodeMetaByAddr map[netip.Addr]nodeMeta, stableConns map[netip.Addr][2]io.ReadWriteCloser) ([]result, error) {
```
Comment (KevinLiang10) [inline cmd/stunstamp/stunstamp.go:244]: Can we have a documentation for this long function here?

File: cmd/stunstamp/stunstamp.go [hunk 5]
```diff
@@ -0,0 +1,778 @@
...
+	log.Println("stunstamp started")
+
+	// Re-using sockets means we get the same 5-tuple across runs. This results
+	// in a higher probability of the packets traversing the same underlay path.
+	// Comparison of stable and unstable 5-tuple results can shed light on
+	// differences between paths where hashing (multipathing/load balancing)
+	// comes into play.
+	stableConns := make(map[netip.Addr][2]io.ReadWriteCloser)
+
+	derpMapTicker := time.NewTimer(time.Minute * 5)
```
Comment (KevinLiang10) [inline cmd/stunstamp/stunstamp.go:675]: Should it be time.NewTicker instead of time.NewTimer?

**Review by KevinLiang10 — COMMENTED** [review]
Everything else LGTM!

File: cmd/stunstamp/stunstamp.go (hunk 5 above)
Comment (jwhited) [reply cmd/stunstamp/stunstamp.go]: updated in https://github.com/tailscale/tailscale/pull/12241/commits/922db653a58b11ad51df33ae45ab29b0c9f3967c

File: cmd/stunstamp/stunstamp.go (hunk 4 above)
Comment (jwhited) [reply cmd/stunstamp/stunstamp.go:248]: updated in https://github.com/tailscale/tailscale/pull/12241/commits/bcc2e04e74e5141fb174824e502680ceb178e77f

**Pull request merged**